colour/io/luts/sony_spimtx.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+"""
+Sony .spimtx LUT Format Input / Output Utilities
+================================================
+
+Defines *Sony* *.spimtx* *LUT* format related input / output utilities objects.
+
+-   :func:`colour.io.read_LUT_SonySPImtx`
+-   :func:`colour.io.write_LUT_SonySPImtx`
+"""
+
+import numpy as np
+
+from colour.constants import DEFAULT_FLOAT_DTYPE
+from colour.io.luts.common import path_to_title
+from colour.io.luts import Matrix
+
+__author__ = 'Colour Developers'
+__copyright__ = 'Copyright (C) 2013-2021 - Colour Developers'
+__license__ = 'New BSD License - https://opensource.org/licenses/BSD-3-Clause'
+__maintainer__ = 'Colour Developers'
+__email__ = 'colour-developers@colour-science.org'
+__status__ = 'Production'
+
+__all__ = ['read_LUT_SonySPImtx', 'write_LUT_SonySPImtx']
+
+
+def read_LUT_SonySPImtx(path):
+    """
+    Reads given *Sony* *.spimtx* *LUT* file.
+
+    Parameters
+    ----------
+    path : unicode
+        *LUT* path.
+
+    Returns
+    -------
+    Matrix
+        :class:`colour.io.Matrix` class instance.
+
+    Examples
+    --------
+    >>> import os
+    >>> path = os.path.join(
+    ...     os.path.dirname(__file__), 'tests', 'resources', 'sony_spimtx',
+    ...     'dt.spimtx')
+    >>> print(read_LUT_SonySPImtx(path))
+    Matrix - dt
+    -----------
+    <BLANKLINE>
+    Dimensions : (3, 4)
+    Matrix     : [[ 0.864274  0.        0.        0.      ]
+                  [ 0.        0.864274  0.        0.      ]
+                  [ 0.        0.        0.864274  0.      ]]
+    Offset     : [ 0.  0.  0.]
+    """
+
+    matrix = np.loadtxt(path, dtype=DEFAULT_FLOAT_DTYPE)
+    matrix = matrix.reshape(3, 4)
+
+    title = path_to_title(path)
+
+    return Matrix(matrix, name=title)
+
+
+def write_LUT_SonySPImtx(matrix, path, decimals=7):
+    """
+    Writes given *LUT* to given *Sony* *.spimtx* *LUT* file.
+
+    Parameters
+    ----------
+    matrix : Matrix
+        :class:`Matrix` class instance to write at given path.
+    path : unicode
+        *LUT* path.
+    decimals : int, optional
+        Formatting decimals.
+
+    Returns
+    -------
+    bool
+        Definition success.
+
+    Examples
+    --------
+    >>> matrix = np.array([[ 1.45143932, -0.23651075, -0.21492857],
+    ...                    [-0.07655377,  1.1762297 , -0.09967593],
+    ...                    [ 0.00831615, -0.00603245,  0.9977163 ]])
+    >>> M = Matrix(matrix)
+    >>> write_LUT_SonySPI1D(M, 'My_LUT.spimtx')  # doctest: +SKIP
+    """
+
+    if matrix.matrix.shape == (3, 4):
+        matrix = matrix.matrix
```

Comment:
Per #792, we should scale the offset by 65535. (The OCIO spimtx reader doesn't seem to mind if the actual values are cast as floats)

Reply:
Thx, I checked the OCIO parser after opening the PR and it raised some more questions, so I stashed for today and opened the PR nonetheless.

Here is some context:

- For SPImtx, the offsets are currently in the last column.
- The new `Matrix` class has separate matrix + offset, i.e. `Matrix.matrix` and `Matrix.offset`.
  - If the matrix is 3x4, the last column acts on a pseudo-channel in the current implementation, i.e. a new white channel concatenated.
    - **Issue**: We never handle alpha in Colour, but for that particular case, we add an alpha channel if the `Matrix.matrix` is 3x4 but we might also probably handle the case where RGB might actually be RGBA, i.e. (..., 4).
    - OCIO applies the 4x4 matrix on RGBA also.
- **Issue**: With that in mind, we simply cannot read the SPImtx matrix directly and shove it in `Matrix.matrix` because then the offset will apply on RGB(**A**) thus we should extract the offset, i.e. last column and put it in  `Matrix.offset`.

Thoughts?

Reply:
Relevant bit for discussion: https://opencolorio.readthedocs.io/en/latest/api/transforms.html#matrixtransform

![image](https://user-images.githubusercontent.com/99779/139498192-2ff90ba3-69cd-45ac-9370-290083a0ed8b.png)

Reply:
Ah yeah -- I think spimtx only encodes twelve values, and none of them are alpha. And when I peeked at the CTL parser, I think I saw something about 5x5 matrices and tried to put it out of my mind. 

Perhaps we should extract the offset and apply as a separate operation on top of the 4x4 or 3x3, input shape pending.

If it's helpful, here's what I'm using to dump PyOpenColorIO.MatrixTransforms to spimtx files:

```
def write_spimtx(xform: ocio.MatrixTransform, path: Union[str, Path], decimals: int = 9):
    matrix, offset = map(as_float_array, (xform.getMatrix(), xform.getOffset()))
    offset *= 65535
    array = np.hstack([np.reshape(matrix, (4, 4))[0:3, 0:3], np.array([offset[0:3]]).T])
    np.savetxt(path, array, fmt='%.{0}f'.format(decimals).encode('utf-8'))
```

Reply:
I will update our code accordingly, what about the `Matrix` class name itself? Any strong opinions?

Reply:
No strong opinions... maybe `MatrixOffset`? (Which we could perhaps one day contrast with `OffsetMatrix`, a la Daniele's Academy Linear Grade proposal :) )

Reply:
Ah good point! I was referring to the OP where suggested to adopt a longer `LUTOperatorMatrix` name for easier discovery but then `LUTOperatorMatrixOffset`? :D

Reply:
Ha -- duh, the OP. Of course. 

`LUTOperatorMatrix` with an argument for applying the offset first sounds great. And I guess for writing spimtxs when the offset is applied first, we could either throw or... offer to write two spimtxs or something :)

Reply:
Just when we thought it would be easy...

![image](https://user-images.githubusercontent.com/99779/139504837-e85caf86-5b48-4690-82c4-6f30457a8f83.png)

Reply:
If we have the argument to the `LUTOperatorMatrix.apply` method it simplifies things a bit, I will put it there for now as it also separates concerns a bit.

Reply:
That sounds perfect. 👍 👍 👍

Reply:
Think we tick the boxes with https://github.com/colour-science/colour/pull/890/commits/d02ae76273ed34dfd09d50129dcda8bfc09f0422